examples/solvers/chambolle_pock_preconditioner_denoising.py

```diff
@@ -0,0 +1,180 @@
+
+   G(x) = {0 if x >=0, infinity if x < 0} ,
+
+respectively. Here, g denotes the image to denoise, ||.||_2 the l2-norm,
+||.||_1 the l1-semi-norm, grad the spatial gradient, lam the regularization
+parameter, |.| the point-wise magnitude across the vector components of
+grad(x), and K is a column vector of operators K = (id, grad)^T with
+identity mapping id.
+
+In order to use the Chambolle-Pock solver, we have to create the column
+operator K, choose a starting point x, create the proximal operator for G,
+create the proximal operator for the convex conjugate of F, choose the
+step sizes tau and sigma such that tau sigma ||K||_2^2 < 1, and set the
+total number of iterations.
+
+For details see :ref:`chambolle_pock`, :ref:`proximal_operators`, and
+references therein.
+"""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+import numpy as np
+import scipy
+import scipy.ndimage
+import matplotlib.pyplot as plt
+
+import odl
+
+# Read test image: use only every second pixel, convert integer to float,
+# and rotate to get the image upright
+image = np.rot90(scipy.misc.ascent()[::2, ::2], 3).astype('float')
+shape = image.shape
+
+# Rescale max to 1
+image /= image.max()
+
+# Discretized spaces
+discr_space = odl.uniform_discr([0, 0], [1, 1], shape)
+
+# Original image
+orig = discr_space.element(image)
+orig.show('Original image')
+
+# Add noise
+image += np.random.normal(0, 0.1, shape)
+
+# Data of noisy image
+noisy = discr_space.element(image)
+noisy.show('Nosy image')
+
+# Gradient operator
+gradient = odl.Gradient(discr_space, method='forward')
+
+# Matrix of operators
+op = odl.BroadcastOperator(odl.IdentityOperator(discr_space), gradient)
+
+# Proximal operators related to the dual variable
+
+# l2-data matching
+prox_convconj_l2 = odl.solvers.proximal_convexconjugate_l2(discr_space,
+                                                           lam=1, g=noisy)
+
+# TV-regularization: l1-semi norm of grad(x)
+prox_convconj_l1 = odl.solvers.proximal_convexconjugate_l1(gradient.range,
+                                                           lam=1/4000.0)
+
+# Combine proximal operators: the order must match the order of operators in K
+proximal_dual = odl.solvers.combine_proximals([prox_convconj_l2,
+                                               prox_convconj_l1])
+
+# Proximal operator related to the primal variable
+
+# Non-negativity constraint
+proximal_primal = odl.solvers.proximal_nonnegativity(op.domain)
+
+
+# --- Run algorithms without preconditioner
+
+
+op_norm_identity = 1.0
+op_norm_gradient = 1.5 * odl.power_method_opnorm(gradient, 100, noisy)
+print('Operator norms, I: {}, gradient: {}'.format(op_norm_identity,
+                                                   op_norm_gradient))
+
+# --- Run algorithms without preconditioner
+
+niter = 50
+
+# Create a function to save the partial errors
```

Comment:
I think it's totally okay to allow functions as callbacks in an algorithm and just define a way it's called. Apart from that, a `ResidualNormPartial` would be an option since it's one of the most common cases.